### PR TITLE
8291911: java/io/File/GetXSpace.java fails with "53687091200 != 161051996160"

### PR DESCRIPTION
**Pull request by bplb**
On Windows, suppress failure if the total space indicated by `df` is less than `FileStore::getTotalSpace` and the free space indicated by `df` equals `FileStore::getUnallocatedSpace`.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8291911](https://bugs.openjdk.org/browse/JDK-8291911): java/io/File/GetXSpace.java fails with "53687091200 != 161051996160"

### Reviewers
 * [Roger Riggs](https://openjdk.org/census#rriggs) (@RogerRiggs - **Reviewer**) ⚠️ Review applies to [81296309](https://git.openjdk.org/jdk/pull/9856/files/8129630943ae98e79bceff18c20a5cb71d30ffd9)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/9856/head:pull/9856` \
`$ git checkout pull/9856`

Update a local copy of the PR: \
`$ git checkout pull/9856` \
`$ git pull https://git.openjdk.org/jdk pull/9856/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 9856`

View PR using the GUI difftool: \
`$ git pr show -t 9856`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/9856.diff">https://git.openjdk.org/jdk/pull/9856.diff</a>

</details>

### CODE REVIEW

**Comment by bplb** [comment]
Passed on 20 arbitrary nodes and on one where it failed previously due to user disk quotas.

File: test/jdk/java/io/File/GetXSpace.java [hunk 1]
```diff
@@ -238,7 +238,12 @@ private static void compare(Space s) {
             // value when the number of blocks is odd.
             if (!Platform.isOSX() || blockSize != 512 || numBlocks % 2 == 0
                 || ts - s.total() != 512) {
-                fail(s.name(), s.total(), "!=", ts);
+                // If Windows user quotas are in effect, then the total space
+                // reported by df might be less that the actual total space in
+                // the store, but the free space should be the same.
+                if (!(Platform.isWindows() && ts > s.total && fs == s.free())) {
```
Comment (bplb) [inline test/jdk/java/io/File/GetXSpace.java:244]: Unsure of the conditional here, in particular whether `fs == s.free()` is needed. Also, when failure is suppressed here, an informational print might be useful.

**Comment by AlanBateman** [comment]
This is hard to review. Would it be possible to provide a summary on the APIs and df behave with quotas?

**Comment by bplb** [comment]
> This is hard to review. Would it be possible to provide a summary on the APIs and df behave with quotas?

We use the Windows API [GetDiskFreeSpaceEx](https://docs.microsoft.com/en-us/windows/win32/api/fileapi/nf-fileapi-getdiskfreespaceexw) wherein it is stated
- lpFreeBytesAvailableToCaller: If per-user quotas are being used, this value may be less than the total number of free bytes on a disk.
- lpTotalNumberOfBytes: If per-user quotas are being used, this value may be less than the total number of bytes on a disk.

The `df` command on the other hand appears to be using something like [GetQuotaInformation](IDiskQuotaUser::GetQuotaInformation) which fills a [structure](https://docs.microsoft.com/en-us/windows/win32/api/dskquota/ns-dskquota-diskquota_user_information) which accounts for the quota and provides the currently used and maximum number of bytes for a user.

**Comment by bplb** [comment]
In reality GetFreeDiskSpaceEx does not appear to recognize the quotas, at least in our tests.

File: test/jdk/java/io/File/GetXSpace.java (hunk 1 above)
Comment (bplb) [reply test/jdk/java/io/File/GetXSpace.java]: Changed the conditional in 991dc9a6bf86e3ec54fd1c244c655a86506a697d.

**Comment by bplb** [comment]
The version of the test in commit 991dc9a6bf86e3ec54fd1c244c655a86506a697d passed 5 repeats on Linux and macOS and 51 repeats on Windows, including on a node on which it previously failed.

**Comment by bplb** [comment]
continue;

**Comment by bplb** [comment]
Commit 8129630943ae98e79bceff18c20a5cb71d30ffd9 passed 50 repeats on the Unix platforms and 150 on Windows.

**Comment by bplb** [comment]
Ping.

**Review by RogerRiggs — APPROVED** [review]
Looks fine.

File: test/jdk/java/io/File/GetXSpace.java [hunk 2]
```diff
@@ -103,27 +103,50 @@ private static class Space {
...
+        }
+    }
+
+    private static void diskFree() throws IOException {
+        ArrayList<Space> al = new ArrayList<>();
+
+        String cmd = "fsutil volume diskFree C:\\";
+        StringBuilder sb = new StringBuilder();
+        Process p = Runtime.getRuntime().exec(cmd);
+        try (BufferedReader in = new BufferedReader(new InputStreamReader(p.getInputStream()))) {
```
Comment (RogerRiggs) [inline test/jdk/java/io/File/GetXSpace.java:140]: The RHS could be replaced by `p.inputReader()`.  (new in JDK 17)

Comment (bplb) [reply test/jdk/java/io/File/GetXSpace.java]: So changed in 02556f265e1fc5274e73e8cfa75ce218c130112e.

@RogerRiggs Thanks for the review.

**Pull request closed without merging**